### PR TITLE
[FIX #1048] show localized timestamp in chat

### LINKED ISSUE #1048: Timestamp format in chat
### Description
image

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Bug

#### Expected behavior
[comment]: # (Describe what you expected to happen.)
Timestamp should be understandable to the user

#### Actual behavior
[comment]: # (Describe what actually happened.)
Timestamp is shown in period time format, but without AM/PM

### PR DESCRIPTION
**Pull request by afhammad**
Fixes #1048 

### Summary:

Timestamps in chat were not localized. See #1048 for more info.

### Steps to test:
See https://github.com/status-im/status-react/issues/1048#issuecomment-345337106

status: ready

### CODE REVIEW

**Comment by pablanopete** [comment]
@afhammad Thanks for your contribution to Status! We'll get someone to review your submission shortly - in the mean time make sure to join our chat at https://chat.status.im/#/register and shoot me a message @hutch so we can make proper introductions. 🎉

File: src/status_im/utils/datetime.cljs [hunk 1]
```diff
@@ -25,12 +25,16 @@
...
+         local       (plus date time-zone-offset)
+         today       (t/today-at-midnight)
+         yesterday   (plus today (days -1))
+         format-opts (clj->js {:day "numeric" 
+                               :month "short" 
+                               :hour "numeric" 
+                               :minute "numeric"})]
      (cond
-       (before? date yesterday) (unparse (formatter "dd MMM hh:mm") local)
+       (before? date yesterday) (.toLocaleString (js/Date. local) locale format-opts)
```
Comment (afhammad) [inline src/status_im/utils/datetime.cljs]: If you think `.toLocaleString` is the right approach to localizing time stamps it might be worth moving it out into a function to be reused elsewhere?, ex: 
```clojure
(defn localize-timestamp
  ([timestamp] (localize-timestamp timestamp {}))
  ([timestamp opts] (.toLocaleString timestamp locale (clj->js opts)))
```

**Comment by afhammad** [comment]
`toLocaleString` works great on iOS but seems to ignore locale on Android. Doesn't seem like its in the pipeline either: https://github.com/facebook/react-native/issues/8349#issuecomment-317562374

Trying out alternative approaches..

File: src/status_im/utils/datetime.cljs [hunk 2]
```diff
@@ -25,12 +25,16 @@
...
   ([ms today-format-fn]
-   (let [date      (from-long ms)
-         local     (plus date time-zone-offset)
-         today     (t/today-at-midnight)
-         yesterday (plus today (days -1))]
+   (let [date        (from-long ms)
+         local       (plus date time-zone-offset)
+         today       (t/today-at-midnight)
+         yesterday   (plus today (days -1))
+         format-opts (clj->js {:day "numeric"
```
Comment (oskarth) [inline src/status_im/utils/datetime.cljs]: Align map values please

**Comment by oskarth** [comment]
Thanks for looking into this @afhammad! Moving this back to in progress until a solution is found for Android.

File: src/status_im/utils/datetime.cljs (hunk 2 above)
Comment (afhammad) [reply src/status_im/utils/datetime.cljs]: Thanks, it may get yanked out depending on Android fix, otherwise will clean it up.

**Comment by afhammad** [comment]
Refactored to use `i18n` library's `localize` method. This requires configuring the formats and data/time related translations in the translation files, more reliable and consistent with the codebase. 

I setup `en` and `de` for testing purposes, also made sure the `localize` wrapper falls back to `en` format if the requested format is not found in the current locale, which means merging this as is should not be worse off than what we have now and we can gradually add more translations.

Tested on iOS and Android sims.

File: env/dev/env/ios/main.cljs [hunk 3]
```diff
@@ -13,7 +13,7 @@
 (def root-el (r/as-element [reloader]))
 
 (figwheel/watch-and-reload
- :websocket-url "ws://localhost:3449/figwheel-ws"
+ :websocket-url "ws://10.0.1.15:3449/figwheel-ws"
```
Comment (jeluard) [inline env/dev/env/ios/main.cljs]: Should not be part of the PR.

File: src/status_im/translations/en.cljs [hunk 4]
```diff
@@ -76,6 +76,19 @@
    :datetime-yesterday                    "yesterday"
    :datetime-today                        "today"
 
+   :date                                  {:formats {:default "%d/%m/%Y"
+                                                     :short   "%d %b"
+                                                     :long    "%A, %d %B %Y"}
+                                            :day-names        ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]
```
Comment (jeluard) [inline src/status_im/translations/en.cljs:82]: Do we want to have to manage this ourselves? Are they no library handling this?

**Review by jeluard — CHANGES_REQUESTED** [review]

**Comment by oskarth** [comment]
@afhammad Any luck with this one? Also see @jeluard's comment.

File: src/status_im/translations/en.cljs (hunk 4 above)
Comment (afhammad) [reply src/status_im/translations/en.cljs:82]: @jeluard Yeh there are other library options. I just opted for using the i18n library already included for consistency however unfortunately it doesn't seem to have a data pack with date/time locales. 

Other options:
1) Google Closure's `goog.i18n.DateTimeFormat`: 
The issue with this is that google closure loads the appropriate locale translations based on `goog.LOCALE` at compile time. See: https://github.com/google/closure-library/blob/master/closure/goog/i18n/datetimesymbols.js#L3423-L3425. However it seems it is possible to manually override it at runtime, although i don't know how reliable that is.
2) Moment.js (https://github.com/cljsjs/packages/tree/master/moment).
3) Do you know of any others?

Happy to try out any of the above

**Comment by afhammad** [comment]
@oskarth responded to @jeluard with some alternatives approaches. Will wait on input from you guys on your preferred direction.

File: src/status_im/i18n.cljs [hunk 5]
```diff
@@ -162,16 +165,29 @@
     (.p rn-dependencies/i18n count (name path) (clj->js options))
     (name path)))
 
+(defn lookup [scope]
+  (.lookup rn-dependencies/i18n scope))
+
+(defn localize
+  ([scope value] (localize scope value {}))
+  ([scope value options]
+   (let [l #(.l rn-dependencies/i18n scope value (clj->js options))]
```
Comment (jeluard) [inline src/status_im/i18n.cljs:174]: Can we extract this as dedicated function with proper default?

File: src/status_im/translations/en.cljs (hunk 4 above)
Comment (jeluard) [reply src/status_im/translations/en.cljs:82]: `moment.js` might be a good alternative. It is a too big burden to have to manage this ourselves.

**Review by jeluard — CHANGES_REQUESTED** [review]

File: src/status_im/translations/en.cljs (hunk 4 above)
Comment (afhammad) [reply src/status_im/translations/en.cljs:82]: ok will try it out

**Comment by oskarth** [comment]
Closing PR due to inactivity. Feel free to reopen this if you feel this was in error.
